libraries/lib-basic-ui/BasicUI.h

```diff
@@ -315,20 +315,24 @@ inline std::unique_ptr<GenericProgressDialog> MakeGenericProgress(
       return nullptr;
 }
 
-template <typename RangeItemType, typename RangeType>
+template <typename ItType>
 void SplitProgress(
-   ProgressReporter parent, RangeType range,
-   std::function<void(RangeItemType, const ProgressReporter&)> action)
+   ItType first, ItType last,
+   std::function<void(typename ItType::value_type, const ProgressReporter&)>
+      action,
+   ProgressReporter parent)
 {
    auto count = 0;
+   const auto numIters = std::distance(first, last);
+   if (numIters == 0)
+      return;
    const ProgressReporter child =
-      parent ?
-         [&](double progress) { parent((count + progress) / range.size()); } :
-         ProgressReporter {};
+      parent ? [&](double progress) { parent((count + progress) / numIters); } :
+               ProgressReporter {};
 
-   for (auto entry : range)
+   for (; first != last; ++first)
    {
-      action(entry, child);
+      action(*first, child);
       ++count;
    }
 }
```

Comment:
Now the utility function is nicer to use, but I feel that BasicUI.h is the wrong place for it.  I'm not sure what is the right place for it.

Reply:
The loops and nested loops surely look much nicer now.

I think the function needs a better name than `SplitProgress`.  Perhaps `WithRedenominatedProgress` and some comments explaining the change of denominator as the size of the given iterator range.

Reply:
I made several attempts at finding a better name, but then came to think that it wasn't such a bad name after all.
Added a description in 1bc93b7.